### DISCUSSION TITLE
Can snmp_exporter read metrics from UPS Schneider ?

### DISCUSSION THREAD

**Discussion by Aguna78** [prometheus/snmp_exporter#1174] [Q&A]
I have Schneider UPS, Easy UPS 3S 40Kva and I want to Monitoring it with Prometheus and Grafana but, there are problem when I get the metrics from Prometheus, the data not suitable with Grafana SNMP dashboard.

### Host operating system: output of `uname -a`
Ubuntu Server 22.04
### snmp_exporter version: output of `snmp_exporter -version`
0.25.0

### What device/snmpwalk OID are you using?
apcups from snmp.yml
### What did you do that produced an error?
# HELP ifAdminStatus The desired state of the interface - 1.3.6.1.2.1.2.2.1.7
# TYPE ifAdminStatus gauge
ifAdminStatus{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1
ifAdminStatus{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1
ifAdminStatus{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 2
# HELP ifConnectorPresent This object has the value 'true(1)' if the interface sublayer has a physical connector and the value 'false(2)' otherwise. - 1.3.6.1.2.1.31.1.1.1.17
# TYPE ifConnectorPresent gauge
ifConnectorPresent{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1
ifConnectorPresent{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 2
ifConnectorPresent{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 1
# HELP ifCounterDiscontinuityTime The value of sysUpTime on the most recent occasion at which any one or more of this interface's counters suffered a discontinuity - 1.3.6.1.2.1.31.1.1.1.19
# TYPE ifCounterDiscontinuityTime gauge
ifCounterDiscontinuityTime{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifCounterDiscontinuityTime{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifCounterDiscontinuityTime{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCInBroadcastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were addressed to a broadcast address at this sub-layer - 1.3.6.1.2.1.31.1.1.1.9
# TYPE ifHCInBroadcastPkts counter
ifHCInBroadcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifHCInBroadcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifHCInBroadcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCInMulticastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were addressed to a multicast address at this sub-layer - 1.3.6.1.2.1.31.1.1.1.8
# TYPE ifHCInMulticastPkts counter
ifHCInMulticastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifHCInMulticastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifHCInMulticastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCInOctets The total number of octets received on the interface, including framing characters - 1.3.6.1.2.1.31.1.1.1.6
# TYPE ifHCInOctets counter
ifHCInOctets{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 5.05995201e+08
ifHCInOctets{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1.24240159e+08
ifHCInOctets{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCInUcastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were not addressed to a multicast or broadcast address at this sub-layer - 1.3.6.1.2.1.31.1.1.1.7
# TYPE ifHCInUcastPkts counter
ifHCInUcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 4.268602e+06
ifHCInUcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 150157
ifHCInUcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCOutBroadcastPkts The total number of packets that higher-level protocols requested be transmitted, and which were addressed to a broadcast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.31.1.1.1.13
# TYPE ifHCOutBroadcastPkts counter
ifHCOutBroadcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifHCOutBroadcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifHCOutBroadcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCOutMulticastPkts The total number of packets that higher-level protocols requested be transmitted, and which were addressed to a multicast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.31.1.1.1.12
# TYPE ifHCOutMulticastPkts counter
ifHCOutMulticastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifHCOutMulticastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifHCOutMulticastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCOutOctets The total number of octets transmitted out of the interface, including framing characters - 1.3.6.1.2.1.31.1.1.1.10
# TYPE ifHCOutOctets counter
ifHCOutOctets{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1.49207702e+08
ifHCOutOctets{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1.24240159e+08
ifHCOutOctets{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHCOutUcastPkts The total number of packets that higher-level protocols requested be transmitted, and which were not addressed to a multicast or broadcast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.31.1.1.1.11
# TYPE ifHCOutUcastPkts counter
ifHCOutUcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 230122
ifHCOutUcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 150157
ifHCOutUcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifHighSpeed An estimate of the interface's current bandwidth in units of 1,000,000 bits per second - 1.3.6.1.2.1.31.1.1.1.15
# TYPE ifHighSpeed gauge
ifHighSpeed{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1000
ifHighSpeed{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 10
ifHighSpeed{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInBroadcastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were addressed to a broadcast address at this sub-layer - 1.3.6.1.2.1.31.1.1.1.3
# TYPE ifInBroadcastPkts counter
ifInBroadcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifInBroadcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifInBroadcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInDiscards The number of inbound packets which were chosen to be discarded even though no errors had been detected to prevent their being deliverable to a higher-layer protocol - 1.3.6.1.2.1.2.2.1.13
# TYPE ifInDiscards counter
ifInDiscards{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifInDiscards{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifInDiscards{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInErrors For packet-oriented interfaces, the number of inbound packets that contained errors preventing them from being deliverable to a higher-layer protocol - 1.3.6.1.2.1.2.2.1.14
# TYPE ifInErrors counter
ifInErrors{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifInErrors{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifInErrors{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInMulticastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were addressed to a multicast address at this sub-layer - 1.3.6.1.2.1.31.1.1.1.2
# TYPE ifInMulticastPkts counter
ifInMulticastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifInMulticastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifInMulticastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInNUcastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were addressed to a multicast or broadcast address at this sub-layer - 1.3.6.1.2.1.2.2.1.12
# TYPE ifInNUcastPkts counter
ifInNUcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifInNUcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifInNUcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInOctets The total number of octets received on the interface, including framing characters - 1.3.6.1.2.1.2.2.1.10
# TYPE ifInOctets counter
ifInOctets{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 5.05995201e+08
ifInOctets{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1.24240159e+08
ifInOctets{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInUcastPkts The number of packets, delivered by this sub-layer to a higher (sub-)layer, which were not addressed to a multicast or broadcast address at this sub-layer - 1.3.6.1.2.1.2.2.1.11
# TYPE ifInUcastPkts counter
ifInUcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 4.268602e+06
ifInUcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 150157
ifInUcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifInUnknownProtos For packet-oriented interfaces, the number of packets received via the interface which were discarded because of an unknown or unsupported protocol - 1.3.6.1.2.1.2.2.1.15
# TYPE ifInUnknownProtos counter
ifInUnknownProtos{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifInUnknownProtos{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifInUnknownProtos{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifIndex A unique value, greater than zero, for each interface - 1.3.6.1.2.1.2.2.1.1
# TYPE ifIndex gauge
ifIndex{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 2
ifIndex{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1
ifIndex{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 3
# HELP ifLastChange The value of sysUpTime at the time the interface entered its current operational state - 1.3.6.1.2.1.2.2.1.9
# TYPE ifLastChange gauge
ifLastChange{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifLastChange{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifLastChange{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifMtu The size of the largest packet which can be sent/received on the interface, specified in octets - 1.3.6.1.2.1.2.2.1.4
# TYPE ifMtu gauge
ifMtu{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1500
ifMtu{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 65536
ifMtu{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 1480
# HELP ifNumber The number of network interfaces (regardless of their current state) present on this system. - 1.3.6.1.2.1.2.1
# TYPE ifNumber gauge
ifNumber 3
# HELP ifOperStatus The current operational state of the interface - 1.3.6.1.2.1.2.2.1.8
# TYPE ifOperStatus gauge
ifOperStatus{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1
ifOperStatus{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1
ifOperStatus{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 2
# HELP ifOutBroadcastPkts The total number of packets that higher-level protocols requested be transmitted, and which were addressed to a broadcast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.31.1.1.1.5
# TYPE ifOutBroadcastPkts counter
ifOutBroadcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifOutBroadcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifOutBroadcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutDiscards The number of outbound packets which were chosen to be discarded even though no errors had been detected to prevent their being transmitted - 1.3.6.1.2.1.2.2.1.19
# TYPE ifOutDiscards counter
ifOutDiscards{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifOutDiscards{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifOutDiscards{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutErrors For packet-oriented interfaces, the number of outbound packets that could not be transmitted because of errors - 1.3.6.1.2.1.2.2.1.20
# TYPE ifOutErrors counter
ifOutErrors{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifOutErrors{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifOutErrors{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutMulticastPkts The total number of packets that higher-level protocols requested be transmitted, and which were addressed to a multicast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.31.1.1.1.4
# TYPE ifOutMulticastPkts counter
ifOutMulticastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifOutMulticastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifOutMulticastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutNUcastPkts The total number of packets that higher-level protocols requested be transmitted, and which were addressed to a multicast or broadcast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.2.2.1.18
# TYPE ifOutNUcastPkts counter
ifOutNUcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifOutNUcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifOutNUcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutOctets The total number of octets transmitted out of the interface, including framing characters - 1.3.6.1.2.1.2.2.1.16
# TYPE ifOutOctets counter
ifOutOctets{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1.49207702e+08
ifOutOctets{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1.24240159e+08
ifOutOctets{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutQLen The length of the output packet queue (in packets). - 1.3.6.1.2.1.2.2.1.21
# TYPE ifOutQLen gauge
ifOutQLen{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 0
ifOutQLen{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 0
ifOutQLen{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifOutUcastPkts The total number of packets that higher-level protocols requested be transmitted, and which were not addressed to a multicast or broadcast address at this sub-layer, including those that were discarded or not sent - 1.3.6.1.2.1.2.2.1.17
# TYPE ifOutUcastPkts counter
ifOutUcastPkts{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 230122
ifOutUcastPkts{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 150157
ifOutUcastPkts{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifPhysAddress The interface's address at its protocol sub-layer - 1.3.6.1.2.1.2.2.1.6
# TYPE ifPhysAddress gauge
ifPhysAddress{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0",ifPhysAddress="00:03:EA:11:C7:C7"} 1
ifPhysAddress{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo",ifPhysAddress="00:00:00:00:00:00"} 1
ifPhysAddress{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0",ifPhysAddress="00:00:00:00:00:00"} 1
# HELP ifPromiscuousMode This object has a value of false(2) if this interface only accepts packets/frames that are addressed to this station - 1.3.6.1.2.1.31.1.1.1.16
# TYPE ifPromiscuousMode gauge
ifPromiscuousMode{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 2
ifPromiscuousMode{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 2
ifPromiscuousMode{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 2
# HELP ifSpecific A reference to MIB definitions specific to the particular media being used to realize the interface - 1.3.6.1.2.1.2.2.1.22
# TYPE ifSpecific gauge
ifSpecific{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0",ifSpecific="0.0"} 1
ifSpecific{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo",ifSpecific="0.0"} 1
ifSpecific{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0",ifSpecific="0.0"} 1
# HELP ifSpeed An estimate of the interface's current bandwidth in bits per second - 1.3.6.1.2.1.2.2.1.5
# TYPE ifSpeed gauge
ifSpeed{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0"} 1e+09
ifSpeed{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo"} 1e+07
ifSpeed{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0"} 0
# HELP ifType_info The type of interface - 1.3.6.1.2.1.2.2.1.3 (EnumAsInfo)
# TYPE ifType_info gauge
ifType_info{ifAlias="",ifDescr="eth0",ifIndex="2",ifName="eth0",ifType="ethernetCsmacd"} 1
ifType_info{ifAlias="",ifDescr="lo",ifIndex="1",ifName="lo",ifType="softwareLoopback"} 1
ifType_info{ifAlias="",ifDescr="sit0",ifIndex="3",ifName="sit0",ifType="tunnel"} 1
# HELP snmp_scrape_duration_seconds Total SNMP time scrape took (walk and processing).
# TYPE snmp_scrape_duration_seconds gauge
snmp_scrape_duration_seconds{module="apcups"} 1.145140644
snmp_scrape_duration_seconds{module="if_mib"} 0.124325292
# HELP snmp_scrape_packets_retried Packets retried for get, bulkget, and walk.
# TYPE snmp_scrape_packets_retried gauge
snmp_scrape_packets_retried{module="apcups"} 0
snmp_scrape_packets_retried{module="if_mib"} 0
# HELP snmp_scrape_packets_sent Packets sent for get, bulkget, and walk; including retries.
# TYPE snmp_scrape_packets_sent gauge
snmp_scrape_packets_sent{module="apcups"} 22
snmp_scrape_packets_sent{module="if_mib"} 8
# HELP snmp_scrape_pdus_returned PDUs returned from get, bulkget, and walk.
# TYPE snmp_scrape_pdus_returned gauge
snmp_scrape_pdus_returned{module="apcups"} 0
snmp_scrape_pdus_returned{module="if_mib"} 122
# HELP snmp_scrape_walk_duration_seconds Time SNMP walk/bulkwalk took.
# TYPE snmp_scrape_walk_duration_seconds gauge
snmp_scrape_walk_duration_seconds{module="apcups"} 1.144504371
snmp_scrape_walk_duration_seconds{module="if_mib"} 0.123771652
# HELP sysUpTime The time (in hundredths of a second) since the network management portion of the system was last re-initialized. - 1.3.6.1.2.1.1.3
# TYPE sysUpTime gauge
### What did you expect to see?
![grafana dashboard](https://github.com/prometheus/snmp_exporter/assets/85914654/15a7ad99-cb31-4694-a188-bd77da090e5b)

### What did you see instead?
No data

**Comment by bastischubert** [comment]
hi @Aguna78 

your device most seems to not use the common mibs, as stated here:  https://community.se.com/t5/EcoStruxure-IT-forum/Easy-UPS-SNMP-MIB-File-Extraction-Procedure/td-p/226841  - you'd need to build a custom config with the correct mib file (can be found in the forum thread)

**Comment by Aguna78** [comment]
Thank you for the reply but how to build mib file? if I already have it.

**Comment by bastischubert** [comment]
You'd need to build a custom snmp.yaml using the generator with that mib file. (See docs at https://github.com/prometheus/snmp_exporter/tree/main/generator )
